prov/gni/include/gnix_tags.h

```diff
@@ -0,0 +1,362 @@
+ *
+ * On receipt of a message:
+ *
+ * fastlock_acquire(&ep->tag_lock);
+ * req = _gnix_remove_by_tag(&ep->posted_recvs, msg->tag, 0);
+ * if (!req)
+ *     _gnix_insert_by_tag(&ep->unexpected_recvs, msg->tag, msg->req);
+ * fastlock_release(&ep->tag_lock);
+ *
+ * On post of receive:
+ *
+ * fastlock_acquire(&ep->tag_lock);
+ * tag_req = _gnix_remove_by_tag(&ep->unexpected_recvs,
+ *           req->tag, req->ignore);
+ * if (!tag_req)
+ *     _gnix_insert_by_tag(&ep->posted_recvs, tag, req);
+ * fastlock_release(&ep->tag_lock);
+ *
+ */
+
+#ifndef PROV_GNI_SRC_GNIX_TAGS_H_
+#define PROV_GNI_SRC_GNIX_TAGS_H_
+
+#include <fi.h>
+#include <fi_list.h>
+
+#include "gnix_util.h"
+
+/* enumerations */
+
+/**
+ * Enumeration for determining the underlying data structure for a
+ * tag storage.
+ *
+ * Using auto will choose one of list, hlist or kdtree based on mem_tag_format.
+ */
+enum {
+	GNIX_TAG_AUTOSELECT = 0,//!< GNIX_TAG_AUTOSELECT
+	GNIX_TAG_LIST,          //!< GNIX_TAG_LIST
+	GNIX_TAG_HLIST,         //!< GNIX_TAG_HLIST
+	GNIX_TAG_KDTREE,        //!< GNIX_TAG_KDTREE
+	GNIX_TAG_MAXTYPES,      //!< GNIX_TAG_MAXTYPES
+};
+
+/**
+ * Enumeration for the tag storage states
+ */
+enum {
+	GNIX_TS_STATE_UNINITIALIZED = 0,//!< GNIX_TS_STATE_UNINITIALIZED
+	GNIX_TS_STATE_INITIALIZED,      //!< GNIX_TS_STATE_INITIALIZED
+	GNIX_TS_STATE_DESTROYED,        //!< GNIX_TS_STATE_DESTROYED
+};
+
+/* forward declarations */
+struct gnix_tag_storage;
+struct gnix_fab_req;
+struct gnix_address;
+
+/* structure declarations */
+/**
+ * @brief Function dispatch table for the different types of underlying structures
+ * used in the tag storage.
+ *
+ * @var insert_tag    insert a request into the tag storage
+ * @var remove_tag    remove a request from the tag storage
+ * @var peek_tag      probe tag storage for a specific tag
+ * @var init          performs specific initialization based on underlying
+ *                     data structure
+ * @var fini          performs specific finalization based on underlying
+ *                     data structure
+ */
+struct gnix_tag_storage_ops {
+	int (*insert_tag)(struct gnix_tag_storage *ts, uint64_t tag,
+			struct gnix_fab_req *req);
+	struct gnix_fab_req *(*remove_tag)(struct gnix_tag_storage *ts,
+			uint64_t tag, uint64_t ignore,
+			uint64_t flags, void *context,
+			struct gnix_address *addr);
+	struct gnix_fab_req *(*peek_tag)(struct gnix_tag_storage *ts,
+			uint64_t tag, uint64_t ignore,
+			uint64_t flags, void *context,
+			struct gnix_address *addr);
+	int (*init)(struct gnix_tag_storage *ts);
+	int (*fini)(struct gnix_tag_storage *ts);
+};
+
+/**
+ * @note The sequence and generation numbers will be used in the future for
+ *       optimizing the search with branch and bound.
+ */
+struct gnix_tag_list_element {
+	 /* entry to the next element in the list */
+	struct slist_entry free;
+    /* has element been claimed with FI_CLAIM? */
+	int claimed;
```

Comment:
Non-null context could indicate that a request is claimed to save 32 bits in the gnix_fab_request.